airflow/lineage/entities.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Any, Dict, List, Optional
+
+import attr
+
+
+@attr.s(auto_attribs=True)
+class Dataset:
+    pass
+
+
+@attr.s(auto_attribs=True)
+class File(Dataset):
+    url: str = attr.ib()
+    type_hint: Optional[str] = None
+
+    meta_schema: str = __name__ + '.File'
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class User(Dataset):
+    email: str = attr.ib()
+    first_name: Optional[str] = None
+    last_name: Optional[str] = None
+    meta_schema: str = __name__ + '.User'
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class Tag(Dataset):
+    tag_name: str = attr.ib()
+    meta_schema: str = __name__ + '.Tag'
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class Column(Dataset):
+    name: str = attr.ib()
+    description: Optional[str] = None
+    data_type: str = attr.ib()
+    tags: List[Tag] = []
+    meta_schema: str = __name__ + '.Column'
+
+
+# this is a temporary hack to satisfy mypy. Once
+# https://github.com/python/mypy/issues/6136 is resolved, use
+# `attr.converters.default_if_none(default=False)`
+def default_if_none(arg: Optional[bool]) -> bool:
+    return arg or False
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class Table(Dataset):
```

Comment:
didn't get the full context, but do you know why we put entities definitions here? And how easy to extend for custom entities?

Reply:
Airflow supports a light weight lineage. Lineage is the information that links Dataset A to Dataset B. Sometimes is important to give some hints to the lineage system to know what is is dealing with ie. File(Dataset A) -> Table(Dataset B).

I on purpose made the basic lineage info resemble Amundsen's model very closely. It will allow databuilder to enrich it or you could implement a Amundsen proxy in Airflow that allows it to push the information as soon as it becomes available. A proxy would also allow you to enrich the information for usage *in* Airflow, so pick up extra information and take actions based on those within the operator or do branching on them based on type of example.

Yes it's very easy to extend. See the papermill_operator for a Notebook example.